### PR TITLE
Recursive Subgraph Fix

### PR DESCRIPTION
**Pull request by Nightmask3**
# Purpose of this PR
> Bugfix for: https://fogbugz.unity3d.com/f/cases/1248437/

# Manual Tests
- Opened shader graph test project locally and ensured all tests are passing
- Tried to create a recursive subgraph using steps mentioned in the bug, with fixes it is no longer possible to do so
- Tried to load a recursive subgraph, it doesn't break the loading of all the other subgraphs

# Comments to reviewers
- @marctem you mentioned these two things in the bug details:
> Be nice to fix 2 things here:
> 1. Having a recursive graph in the project should not break every Shader Graph. (Also check if there might be other causes of import failure lurking in that code)
> 2. If we can catch the recursion during the Save As case we should prevent it, or at least give a clear error message about the situation immediately.
- I believe I've addressed both of those cases, now even if a subgraph does fail to load (which was the case with the recursive subgraphs) it will fail more gracefully instead of breaking all the other graphs.

### CODE REVIEW

**Comment by LandonTownsendUnity** [comment]
Shadergraph is now no longer broken by subgraph import errors, recursion or otherwise (I created one by editing the json and changing the output to a "vector5")
Being able to create a recursive subgraph through save as is fixed, however you can still create one using "convert to subgraph". Requesting that that's fixed as well.

File: com.unity.shadergraph/Editor/Drawing/MaterialGraphEditWindow.cs [hunk 1]
```diff
@@ -422,6 +422,16 @@ bool SaveAsImplementation()
 
                 if (newPath != path)
                 {
+                    var subGraphNodes = graphObject.graph.GetNodes<SubGraphNode>();
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Drawing/MaterialGraphEditWindow.cs:425]: I'm concerned that this only catches a single level of recursion. Please see the recursion check in `SubGraphNode.ValidateNode()` for a deeper way to do it. And if that code can be made a static function on SubGraphNode that can be called there and here, even better.

Comment (cdxntchou) [reply com.unity.shadergraph/Editor/Drawing/MaterialGraphEditWindow.cs]: Fixed to use the pre-baked descendents list, which can catch any depth of recursion.

**Comment by cdxntchou** [comment]
Fixed both 'Save As' with a deeper recursion, as well as added the same check for "Convert To SubGraph".

This PR should make it much more difficult to create recursive loops from within the Editor.

It's still possible to just copy them in or manually create them by editing file text, but at least you can't accidentally create them via Save As or Convert To SubGraph.

The Project insta-crash is a trunk/editor issue; asset-database-v2 team are looking into it.
Once they fix that, recursive subgraphs shouldn't be an issue anymore.

**Comment by cdxntchou** [comment]
Closing in favor of new PR https://github.com/Unity-Technologies/Graphics/pull/888

**Pull request closed without merging**